### PR TITLE
Fallback video player support for Code Studio Markdown content

### PR DESCRIPTION
**Pull request by wjordan**
This feature adds fallback video player support to Code Studio Markdown-generated course content.

Summary:
- The [YoutubeRewriter](https://github.com/code-dot-org/code-dot-org/compare/markdown-fallback-videos?expand=1#diff-2e858e523a81672bc9b787dce7640183R10) class extends the existing Code Studio Markdown renderer, adding a filter that rewrites YouTube `<iframe src="https://www.youtube.com/embed/xyz" />`-type embed tags to source `/videos/embed/xyz` instead.
- `embed.html.haml` was updated to include an inline filter that [processes YouTube videos](https://github.com/code-dot-org/code-dot-org/compare/markdown-fallback-videos?expand=1#diff-0d907e3e32d660e462ae53b1fdba4d1eR1) to make them automatically available through the fallback player, without additional content-creator intervention beyond visiting the newly-created page.
  - The filter is only activated by a logged-in admin account in non-production/test environments (e.g. development, staging, levelbuilder). I've also made a change to automatically force the fallback player when these same conditions are true, for more thorough testing purposes. (At least initially- I can revert this additional change if desired)
- I've also added a [standalone script](https://github.com/code-dot-org/code-dot-org/compare/markdown-fallback-videos?expand=1#diff-82584085d2d8fd290a96fbe020e0a42eR1) to invoke the same video-processing code directly from the command-line. I've included a sample shell command as a [comment](https://github.com/code-dot-org/code-dot-org/compare/markdown-fallback-videos?expand=1#diff-82584085d2d8fd290a96fbe020e0a42eR14) that will run the script on all of the videos that currently exist throughout our level content. I've run this command once already, so all of our current level content has already been fully processed for fallback-readiness.

See [spec](http://wiki.code.org/display/PROD/Online+PD+Fallback+Video+Player+spec) for further details&discussion.

### CODE REVIEW

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 1]
```diff
@@ -135,7 +135,9 @@ function setupVideoFallback(videoInfo, playerWidth, playerHeight, shouldStillAdd
   }
 
   if (videoInfo['force_fallback']) {
-    addFallbackVideoPlayer(videoInfo, playerWidth, playerHeight);
+    setTimeout(function() {
+      addFallbackVideoPlayer(videoInfo, playerWidth, playerHeight);
+    }, 0);
```
Comment (breville) [inline dashboard/app/assets/javascripts/videos.js.erb]: comment about why the timeout is necessary?

**Comment by breville** [comment]
can you explain a bit more about who should used embed and when?  similar for the command line i guess?

is this a process that is run again each time new videos are added?

File: dashboard/app/assets/javascripts/videos.js.erb (hunk 1 above)
Comment (wjordan) [inline dashboard/app/assets/javascripts/videos.js.erb]: The added timeouts fix an ordering bug in the `createVideoWithFallback` code path:

`createVideoWithFallback` function returns the newly-created `<iframe id="video"/>` element, which gets appended to the #video-container by the calling script. However, before this function returns `setupVideoFallback` is called, which calls `addFallbackVideoPlayer`, which expects to find the iframe element in the DOM tree. In most cases this worked okay because `onYouTubeBlocked`/`testImageAccess` calls `addFallbackVideoPlayer` asynchronously (so the iframe element existed by the time it was called), but in some cases (e.g., when `fallback_only` or `force_youtube_fallback` are set), the callback would be invoked immediately, causing an error.

Adding these timeouts fixed this error for these edge cases in my manual tests.

**Comment by wjordan** [comment]
The embed view creates a page that transparently provides fallback-player support to the specified YouTube video. (I'll add a comment to embed.html.haml with this brief summary)

It was originally written by Brian a year ago (PR: [dashboard#918](https://github.com/code-dot-org/dashboard/pull/918)) to be used on the video-test for the [/educate/it](https://code.org/educate/it) page. It hasn't seen much use until now which is why a couple bugfixes and tweaks were needed.

The command-line is just a utility script to be made available for future use, there is no process underlying it. Each time new videos are added, they will automatically be processed as soon as a content-creator adds a new iframe to a markdown page and views the rendered result from a logged-in account. (If this is not sufficient, we could add a hook within our CI script to ensure the processing of all videos using the supplied script and command-line, but I wouldn't recommend this unless anyone has issues with the more lightweight automatic processing filter.)

**Comment by sfilman** [comment]
Hey @wjordan I wanted to try to capture the "education/pm team friendly" steps for this to make sure I understand.

It seems like it's as easy as:
1. Existing embed tags across all scripts should be good to go
2. Use the embedded iframe tag as before in an external HTML level, so no change
3. Be sure to view the rendered page w/with the markdown from a logged in admin account in any non-prod environment to trigger fallback player support

Is that right?

Also, is there a good way to verify with existing levels where the embedded youtube has been added already that it will successfully trigger the fallback player? i.e. how do I know that step 1 & 3 "worked"

**Comment by wjordan** [comment]
@sfilman those points sound correct, there should be very little change to the ed/pm-team process. In particular, #3 shouldn't require any extra steps when creating a level through the Code Studio web interface- since a logged-in admin account is already required to create/edit a level via the level builder in the first place, and saving/updating a level automatically redirects the admin to the rendered page. The only exception to this would be when a page is added/edited in Markdown file directly, e.g. via Dropbox sync or direct Git commit (which as far as I know, happens only rarely/not at all).

To help verify the fallback player with existing levels, I've made a change to automatically force the fallback player for logged-in admin accounts on non-prod/test environments. Once this PR gets merged to staging/levelbuilder, it will be possible to test out the fallback-player in those environments.

File: bin/upload-video [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+# Usage:
+# upload-video [youtube id] [video filename]
+#
+# Dependency requirements:
+#
+# ffmpeg (`brew install ffmpeg` / `apt-get install ffmpeg`) - for transcoding manually-provided video files
+# s3_access_key_id and s3_secret_access_key defined (e.g. in `locals.yml`)
+
+# To use this script to process all YouTube video tags in custom levels, run this command from the root folder (between =begin and =end):
```
Comment (philbogle) [inline bin/upload-video]: Would it make sense to check in the script for processing all YouTube video tags in custom levels?  (That can happen in another pull request)

File: bin/upload-video [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+end
+
+def main
+  if ARGV.length < 1
+    return usage
+  end
+  id = ARGV[0]
+  filename = ARGV[1]
+  if id.nil?
+    return usage
```
Comment (philbogle) [inline bin/upload-video]: I think this is easier to read and more obvious if you just split "return usage" across two lines; i.e.

  print_usage
  return

("return usage" led me on a detour to figure out what the return value of usage" was and then pondering whether returning nil from main had any significance.)

File: bin/upload-video [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+require 'cdo/video/youtube'
+
+VIDEO_BUCKET = 'videos.code.org'
+
+def usage
+  puts 'Usage: upload-video [youtube id] [video filename]'
+end
+
+def main
+  if ARGV.length < 1
```
Comment (philbogle) [inline bin/upload-video]: Doesn't this require 2 arguments?

File: bin/upload-video (hunk 2 above)
Comment (wjordan) [inline bin/upload-video]: Do you mean checking in this commented-out command as a standalone shell script, so it could be invoked directly from the command-line as opposed to copying/pasting it to the shell as-needed?

My original thought was that it made sense to bury it in a comment here, as I didn't intend for it to be part of any regular process (it's a one-off script), while at the same time preserving it if it's ever needed again in the future. I don't have strong preference on this though so I can commit it as a separate shell script if that's your preference

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 5]
```diff
@@ -135,7 +135,11 @@ function setupVideoFallback(videoInfo, playerWidth, playerHeight, shouldStillAdd
   }
 
   if (videoInfo['force_fallback']) {
-    addFallbackVideoPlayer(videoInfo, playerWidth, playerHeight);
+    // this and the below setTimeout wrapper work around a load-ordering issue, see detailed comment:
+    // https://github.com/code-dot-org/code-dot-org/pull/3260#discussion_r35707401
+    setTimeout(function() {
+      addFallbackVideoPlayer(videoInfo, playerWidth, playerHeight);
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/videos.js.erb]: I read the detailed comment but still don't fully understand the scenario.

Is it absolutely guaranteed that the iframe element will exist when the timeout function is called, or just very likely?  Is there some event that you can fire on that will guarantee that the needed iframe has been created?

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 6]
```diff
@@ -143,7 +147,9 @@ function setupVideoFallback(videoInfo, playerWidth, playerHeight, shouldStillAdd
     if (!shouldStillAddCallback()) {
       return;
     }
-    addFallbackVideoPlayer(videoInfo, playerWidth, playerHeight);
+    setTimeout(function() {
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/videos.js.erb]: The setTimeout here duplicates the code seen on line 138; perhaps have the logic encapsulated in addFallbackVideoPlayer so that the caller doesn't have to worry about it?

File: dashboard/app/controllers/videos_controller.rb [hunk 7]
```diff
@@ -79,7 +79,14 @@ def set_video
   end
 
   def set_video_by_key
-    @video = Video.find_by_key(params[:key])
+    key = params[:key]
+    # Create a temporary video object from default attributes if an entry isn't found in the DB
```
Comment (philbogle) [inline dashboard/app/controllers/videos_controller.rb]: Trailing period.

File: dashboard/app/controllers/videos_controller.rb [hunk 8]
```diff
@@ -79,7 +79,14 @@ def set_video
...
+    @video = Video.find_by_key(key) ||
+      Video.new(
+        key:key,
```
Comment (philbogle) [inline dashboard/app/controllers/videos_controller.rb]: Space after :

File: dashboard/app/views/videos/embed.html.haml [hunk 9]
```diff
@@ -1,6 +1,14 @@
+/ This view creates a page that transparently provides fallback-player support to the specified YouTube video.
+/ See VideosController#embed for usage
+!!! 5
+- video = local_assigns[:video]
+- process_fallback_videos = current_user.try(:admin?) && !(Rails.env.production? || Rails.env.test?)
```
Comment (philbogle) [inline dashboard/app/views/videos/embed.html.haml]: The conditional is a little tricky to understand.  I think the following is equivalent, but I don't quite understand the intent. I see why you wouldn't want to process fallback videos in test, but it seems like we *do* want to process fallback videos in production. (?)

process_fallback_videos = current_user.try(:admin?) && !Rails.env.production? && !Rails.env.test?

File: dashboard/app/views/videos/embed.html.haml [hunk 10]
```diff
@@ -1,6 +1,14 @@
...
+- Youtube.process(video.key) if process_fallback_videos
```
Comment (philbogle) [inline dashboard/app/views/videos/embed.html.haml]: Could you move the business logic like Youtube video processing into the controller?  This seems too heavyweight for a view and is hard to test properly in that context.

File: dashboard/config/initializers/markdown_handler.rb [hunk 11]
```diff
@@ -6,8 +7,29 @@ module MarkdownHandler
     space_after_headers: true
   }
 
+  # Rewrite youtube iframe elements to use the fallback-player iframe instead
```
Comment (philbogle) [inline dashboard/config/initializers/markdown_handler.rb]: s/youtube/YouTube/

File: dashboard/config/initializers/markdown_handler.rb [hunk 12]
```diff
@@ -6,8 +7,29 @@ module MarkdownHandler
...
+  class YoutubeRewriter < Redcarpet::Render::HTML
+    def block_html(html)
+      doc = ::Nokogiri::HTML(html)
+      nodes = doc.xpath(%w(youtube youtubeeducation).map{|x| "//iframe[@src[contains(.,'#{x}.com/embed')]]"}.join(' | '))
+      nodes.each do |node|
+        if node['src']
+          regex = /(http:|https:)?\/\/(www\.)?(youtube|youtubeeducation)\.com\/embed\/([^"&?\/ ]{11})/
```
Comment (philbogle) [inline dashboard/config/initializers/markdown_handler.rb]: Please explain the magic number 11 (e.g. by making it a constant or computing it somehow.)

File: dashboard/config/initializers/markdown_handler.rb [hunk 13]
```diff
@@ -6,8 +7,29 @@ module MarkdownHandler
...
+          id = node['src'].match(regex)[4]
```
Comment (philbogle) [inline dashboard/config/initializers/markdown_handler.rb]: Consider using named captures to make this more readable and robust (in case the regular expression changes)

http://ruby-doc.org/core-1.9.3/Regexp.html#method-i-named_captures

File: dashboard/config/initializers/markdown_handler.rb [hunk 14]
```diff
@@ -6,8 +7,29 @@ module MarkdownHandler
...
+          node['src'] = node['src'].sub(
```
Comment (philbogle) [inline dashboard/config/initializers/markdown_handler.rb]: I think you can use sub! here.

File: dashboard/config/initializers/markdown_handler.rb [hunk 15]
```diff
@@ -6,8 +7,29 @@ module MarkdownHandler
...
+            regex,
+            Rails.env.development? ?
+              "/videos/embed/#{id}" :
```
Comment (philbogle) [inline dashboard/config/initializers/markdown_handler.rb]: Could you encapsulate the logic for computing in the url_for a video in a helper?  It seems likely we might construct such URLs in multiple places.

File: lib/cdo/video/youtube.rb [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+  # Process a video with a youtube identifier.
+  # If filename is provided, transcode the local file with ffmpeg.
+  # If no filename is provided, download the transcoded video from YouTube.
+  #
+  # When downloading from YouTube, an HTTP head request will first check the absence of the file.
+  # If `force`==true, the head request will be skipped.
+  def self.process(id, filename=nil, force=false)
+    if filename.nil? && !force
+      require 'httparty'
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Could we move this require to the top of the file to make it more obvious what the file depends on?

File: lib/cdo/video/youtube.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+  VIDEO_BUCKET = 'videos.code.org'
+
+  # Process a video with a youtube identifier.
+  # If filename is provided, transcode the local file with ffmpeg.
+  # If no filename is provided, download the transcoded video from YouTube.
+  #
+  # When downloading from YouTube, an HTTP head request will first check the absence of the file.
+  # If `force`==true, the head request will be skipped.
+  def self.process(id, filename=nil, force=false)
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Do we need to do this work in a more asynchronous fashion?  This is called from a web page and seems do a lot of blocking operations; it seems like we might have issues with requests timing out or at least tying up worker threads.

File: lib/cdo/video/youtube.rb [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    if filename.nil? && !force
+      require 'httparty'
+      thumbnail_url = "https://#{VIDEO_BUCKET}/youtube/#{id}.jpg"
+      response = HTTParty.head(thumbnail_url).response
+      if response.is_a? ::Net::HTTPSuccess
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Suppose the video actually does exist but we get a spurious failure.  Is there no harm to re-rendering and re-uploading the video?  (Perhaps not, looking below, just wanted to verify!)

File: lib/cdo/video/youtube.rb [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        puts "Video id: #{id} already processed"
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Should this be using some kind of logger rather than a straight puts?

File: lib/cdo/video/youtube.rb [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+          -crf 22
+          -s 640x360
+        ).join(' ')} #{output_file}"
+      else
+        # Run viddl-rb to download transcoded video from youtube
+        url = "https://www.youtube.com/watch?v=#{id}"
+        cmd = "viddl-rb #{url} -s #{dir} -q 640:360:mp4"
+      end
+
+      IO.popen(cmd) { |output| output.each { |line| puts line } }
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: I'm not sure this will detect all cases when the cmd fails.  (The mp4 file might be created but incomplete for example.)

I think it would be good to check the return code from cmd to make sure it thinks it succeeded.  If there are additional sanity checks you can do that the file is valid and complete, that would also be good.

File: dashboard/app/views/videos/embed.html.haml (hunk 10 above)
Comment (philbogle) [inline dashboard/app/views/videos/embed.html.haml]: This will throw an exception if the youtube video processing fails.  Could you arrange for the user to see a helpful error page that explains what happened and what to do?

File: lib/cdo/video/youtube.rb [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+      file = Dir.glob("#{dir}/*").first
+      raise RuntimeError, 'Video not available in correct format' if File.extname(file) != '.mp4'
+      video_filename = AWS::S3.upload_to_bucket(VIDEO_BUCKET, "youtube/#{id}.mp4", File.open(file), access: :public_read, no_random: true, content_type: 'video/mp4')
+      puts "https://#{VIDEO_BUCKET}/#{video_filename}"
+      thumbnail_file = "https://i.ytimg.com/vi/#{id}/0.jpg"
+      thumbnail = open(thumbnail_file) || raise(RuntimeError, 'Could not retrieve thumbnail for video')
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Please remove the parenthesis after raise for consistency with other places in this file.

File: lib/cdo/video/youtube.rb [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+          -ac 2
+          -ab 128k
+          -vcodec libx264
+          -preset slow
+          -f mp4
+          -crf 22
+          -s 640x360
+        ).join(' ')} #{output_file}"
+      else
+        # Run viddl-rb to download transcoded video from youtube
```
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: What is the testing and monitoring plan for this feature?

I am guessing we can get by with ad-hoc testing initially, but it does seem like we need monitoring to determine whether we are getting lots of errors processing videos in production.

**Comment by philbogle** [comment]
Done with my initial review pass.

File: dashboard/config/initializers/markdown_handler.rb (hunk 14 above)
Comment (wjordan) [inline dashboard/config/initializers/markdown_handler.rb]: I originally used sub! here, but it didn't work so I changed it - it seems that Nokogiri overrides the hash-accessor functions, or something. Here's an `irb` comparison:
```
2.2.1 :021 > Nokogiri::HTML('<iframe src="http://www.youtube.com"/>').tap{|x|x.xpath('//iframe').first['src'].sub!('youtube','X')}.css('body').children.to_html
 => "<iframe src=\"http://www.youtube.com\"></iframe>" 

2.2.1 :022 > Nokogiri::HTML('<iframe src="http://www.youtube.com"/>').tap{|x|x.xpath('//iframe').first.tap{|y|y['src'] = y['src'].sub('youtube','X')}}.css('body').children.to_html
 => "<iframe src=\"http://www.X.com\"></iframe>" 
```

File: bin/upload-video (hunk 4 above)
Comment (wjordan) [inline bin/upload-video]: The intention was to make the second argument optional, in which case it would download the file from YouTube. I've updated the usage and documentation in this script wrapper to make this usage more clear.

File: lib/cdo/video/youtube.rb (hunk 21 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: parentheses can't be removed for method arguments following a binary operator, e.g.:

```
2.2.1 :001 > puts 0
0

2.2.1 :002 > false || puts 0
SyntaxError: (irb):2: syntax error, unexpected tINTEGER, expecting keyword_do or '{' or '('

2.2.1 :003 > false || puts(0)
0
```

File: lib/cdo/video/youtube.rb (hunk 18 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: There shouldn't be any harm if a spurious failure causes the filter to re-process the video. I made this condition intentionally narrow so that the filter would err on the side of caution in the case of uncertainty.

File: lib/cdo/video/youtube.rb (hunk 17 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: Since the processing is only done by logged-in admin accounts on non-production/test environments on newly-added videos, this blocking operation would not pose any critical load-related issues to the live site. The worst that could happen is that if the process blocks for over one minute, the admin gets a 503 timeout page and contacts an engineer to investigate. We could look into rewriting this into an asynchronous fashion at that time if the issue persists.

File: lib/cdo/video/youtube.rb (hunk 22 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: The monitoring plan is that if an author adds a new YouTube iframe tag that produces a processing error or fails to load in their page correctly, then they report the problem to an engineer and we can investigate. Since the fallback player will be force-enabled in levelbuilder and staging environments, the author can ensure that the fallback video has been correctly processed and loads as expected in their browser as soon as the new video tag is added to their content. This will use the same testing and monitoring process as e.g. detecting syntax-parsing errors in regular Markdown content.

File: bin/upload-video (hunk 2 above)
Comment (wjordan) [inline bin/upload-video]: committed this change.

File: bin/upload-video (hunk 3 above)
Comment (wjordan) [inline bin/upload-video]: done

File: dashboard/app/assets/javascripts/videos.js.erb (hunk 5 above)
Comment (wjordan) [inline dashboard/app/assets/javascripts/videos.js.erb]: refactored this a bit to eliminate the need for a timeout and added comments to clarify the expected preconditions

File: lib/cdo/video/youtube.rb (hunk 16 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: done

File: dashboard/config/initializers/markdown_handler.rb (hunk 15 above)
Comment (wjordan) [inline dashboard/config/initializers/markdown_handler.rb]: refactored this a bit (using the `url_for` helper turned out to be more difficult than it was worth, so I've simplified it in `Video#embed_url`).

File: dashboard/config/initializers/markdown_handler.rb (hunk 13 above)
Comment (wjordan) [inline dashboard/config/initializers/markdown_handler.rb]: Thanks for this suggestion! I like this much better than using numbered capture groups.
I've also added `?:` modifiers for the unused non-capturing groups to make the regex results a bit more clear.

File: dashboard/config/initializers/markdown_handler.rb (hunk 12 above)
Comment (wjordan) [inline dashboard/config/initializers/markdown_handler.rb]: done (extracted into constants, and added reference links for the format)

File: lib/cdo/video/youtube.rb (hunk 20 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: added a line to check the return code of the cmd. (Although `viddl-rb` does not seem to set the error code on all failure conditions, in these cases it fails to produce an mp4 file and so the next line will raise an error.)

File: lib/cdo/video/youtube.rb (hunk 19 above)
Comment (wjordan) [inline lib/cdo/video/youtube.rb]: We have a mix of `$log`, `Pegasus.logger`, `Rails.logger`, `slog`, and `puts` in use throughout the codebase, with `puts` often used in small scripts and basic standard output, but no guidelines on which logger to use where and why. Last month, I added a [simple logger implementation](https://github.com/code-dot-org/code-dot-org/commit/058868917f844e7e0ec0331a624c064e20fe0963#diff-63ce6b57dd8496d4e8739cf50dd128e7R179) to `CDO.log`, since we had no other standard-out logger available across our sub-environments for use in non-Pegasus/Rails scripts. This logger's functionality is currently no different from `puts`, but for now I've changed these new lines to use `CDO.log`- I'd love to see someone eventually have time to work on cleaning up/standardizing our logging practices throughout our codebase.

File: dashboard/app/views/videos/embed.html.haml (hunk 10 above)
Comment (wjordan) [inline dashboard/app/views/videos/embed.html.haml]: done (extracted processing logic into controller) and done (added exception-handler error page)

File: dashboard/app/views/videos/embed.html.haml (hunk 9 above)
Comment (wjordan) [inline dashboard/app/views/videos/embed.html.haml]: In this spec + implementation, we explicitly do not want to process fallback videos in production. Fallback videos only ever need to be processed once, which will usually take place when the content is initially added by the author in the levelbuilder environment. Disabling the filter in production isolates concerns between authoring and delivery of this content, eliminating the video-processing dependencies from being installed in production and protecting the live site from any unintended bugs or issues related to timeouts or tying up worker threads you asked about in [another comment](https://github.com/code-dot-org/code-dot-org/pull/3260#discussion_r35794496).

File: lib/cdo/video/youtube.rb (hunk 19 above)
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Ok, will look at this.  It would be very useful if CDO.log at least included timestamps etc. if it doesn't already.

File: lib/cdo/video/youtube.rb (hunk 22 above)
Comment (philbogle) [inline lib/cdo/video/youtube.rb]: Could you please add a logging statement for errors so that we could at least in principle monitor how many things are failing in an automated fashion.  (I realize our CDO.log system probably doesn't enable this but it will at least remind us of a place that needs to be dealt with.)

**Comment by philbogle** [comment]
LGTM after the one additional logging statement suggested in youtube.rb

**Comment by wjordan** [comment]
Added the extra logging to youtube.rb (integrated CDO.log with Rails.logger, which currently appends timestamps in non-development environments). Thanks @breville @sfilman for comments and @philbogle for the thorough code review!
